File: blabla/src/blabla.cpp
#include "blabla.h"
#include <cstring>
#include <string>
#include <test_repo/sample_library.hpp>

FFI_PLUGIN_EXPORT int sum(int a, int b) { return test_repo::add(a, b); }
FFI_PLUGIN_EXPORT int factorial(int input) {
  return test_repo::factorial(input);
}
FFI_PLUGIN_EXPORT const char *getVersion() {
  return strdup(test_repo::getProjectVersion().c_str());
}
FFI_PLUGIN_EXPORT void process_image(uint8_t *input_data, const int width,
                                     const int height, const int channels,
                                     uint8_t *output_data) {
  test_repo::process_image(input_data, width, height, channels, output_data);
}

File: blabla/lib/blabla.dart
import 'dart:ffi';
import 'dart:io';
import 'dart:ui';
import 'dart:math';
import 'package:ffi/ffi.dart';

import 'blabla_bindings_generated.dart';

const String _libName = 'blabla';

/// The dynamic library in which the symbols for [blablaBindings] can be found.
final DynamicLibrary _dylib = () {
  if (Platform.isMacOS || Platform.isIOS) {
    return DynamicLibrary.open('$_libName.framework/$_libName');
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError('Unknown platform: ${Platform.operatingSystem}');
}();

/// The bindings to the native functions in [_dylib].
final blablaBindings _bindings = blablaBindings(_dylib);

final class SampleLib {
  const SampleLib._();

  static int sum(int a, int b) => _bindings.sum(a, b);
  static String getVersion() =>
      _bindings.getVersion().cast<Utf8>().toDartString();
  static int factorial(int input) => _bindings.factorial(input);
  static void processImage() {
    final width = 5;
    final height = 5;
    final channels = 4; // RGBA

    final inputPointer = malloc<Uint8>(width * height * channels);
    final outputPointer = malloc<Uint8>(width * height * channels);

    // Copy input bytes to native memory
    for (var i = 0; i < width * height * channels; i++) {
      inputPointer[i] = Random().nextInt(256);
    }

    // Process the image
    _bindings.process_image(
      inputPointer,
      width,
      height,
      channels,
      outputPointer,
    );

    // Free native memory
    malloc.free(inputPointer);
    malloc.free(outputPointer);
  }
}
